### PR TITLE
[WIP] Fix search() when mount point overlaps an existing folder

### PR DESCRIPTION
**Pull request by PVince81**
Currently it's still possible to make a system mount point overlap user folders, so we need a workaround to avoid searching into these hidden/overlapped folders.

In the spirit of "test first", this PR first provides unit tests to cover this case.
To make the matter worse: there are two places in the code that implement a similar logic (duplicate code!)

Fix for https://github.com/owncloud/core/issues/13096#issuecomment-70246172
- [x] actual fix
- [x] unit test

To fix this I'll probably have to iterate over the search results and delete the ones that match an existing mount point. Might not be good for performance.

CC @MorrisJobke @icewind1991

### CODE REVIEW

**Comment by karlitschek** [comment]
:+1:

**Comment by PVince81** [comment]
I give up trying to refactor Node API + View's search into one (WIP here: https://github.com/owncloud/core/pull/13420)

This means the fix will need to be put in two separate places

**Comment by PVince81** [comment]
Here we go, I added an extra check in both locations. It is likely to kill performance a bit when searching through big folders because for every result we need to check whether it's from a hidden folder.

I'm not sure whether fixing this is even a good idea... ideally it shouldn't be possible to cover existing folders when mounting storages.

@icewind1991 please have a look and tell me this is a bad fix and that it indeed kills performance...
Another idea would be to pass a list of excluded folder to the cache's "search()" methods, and exclude using a `WHERE` clause. But it all sounds like overkill.

**Comment by PVince81** [comment]
I haven't tested this extensively yet because I'd like to discuss the solution first.
(but it does fix https://github.com/owncloud/core/issues/13096#issuecomment-70246172)

File: lib/private/files/node/folder.php [hunk 1]
```diff
@@ -264,30 +264,46 @@ private function searchCommon($method, $args) {
...
-
-				$relativeMountPoint = substr($mount->getMountPoint(), $rootLength);
-				$results = call_user_func_array(array($cache, $method), $args);
-				foreach ($results as $result) {
-					$result['internalPath'] = $result['path'];
-					$result['path'] = $relativeMountPoint . $result['path'];
-					$result['storage'] = $storage;
-					$files[] = $result;
-				}
+		foreach ($mountPoints as list($subStorage, $relativeMountPoint)) {
```
Comment (icewind1991) [inline lib/private/files/node/folder.php]: `as list(...)` only work in php5.5+

Comment (PVince81) [inline lib/private/files/node/folder.php]: Ahh.... for once that I use fancy stuff :wink:

**Comment by PVince81** [comment]
Also note that this doesn't cover the crazy cases where ext storage 1 is mounted onto a subfolder of ext storage 2... that would be even worse for performance.

**Comment by PVince81** [comment]
I don't like this approach, introduces performance issues to only fix a corner case.

Closing.

**Pull request closed without merging**